File: 剑指Offer/26.树的子结构.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def doesTree1HasTree2(self, A: TreeNode, B: TreeNode) -> bool:
        if not B: return True
        if not A: return False
        if A.val != B.val: return False

        return self.doesTree1HasTree2(A.left, B.left) and self.doesTree1HasTree2(A.right, B.right)

    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        result = False

        if A and B:
            if A.val == B.val:
                result = self.doesTree1HasTree2(A, B)
            if result != True:
                result = self.isSubStructure(A.left, B)
            if result != True:
                result = self.isSubStructure(A.right, B)
        return result

File: 剑指Offer/26.树的子结构.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    bool doesTree1HasTree2 (TreeNode* A, TreeNode* B) {
        if (B == NULL) return true;
        if (A == NULL) return false;
        if (A->val != B->val) return false;
        return doesTree1HasTree2(A->left, B->left) && doesTree1HasTree2(A->right, B->right);
    }
public:
    bool isSubStructure(TreeNode* A, TreeNode* B) {
        bool result = false;

        if (A != NULL && B != NULL) {
            if (A->val == B->val) {
                result = doesTree1HasTree2(A, B);
            }
            if (!result) {
                result = isSubStructure(A->left, B);
            }
            if (!result) {
                result = isSubStructure(A->right, B);
            }
        }
        return result;
    }
};